File: src-tauri/src/fs/moving.rs
use std::{
    collections::{HashMap, VecDeque},
    path::{Path, PathBuf},
    sync::Arc,
};

use tokio::{
    fs,
    io::{self},
    sync::Mutex,
};

use futures::stream::{self, StreamExt as FuturesStreamExt, TryStreamExt};

use crate::bms::{BMS_FILE_EXTS, BMSON_FILE_EXTS};

use super::{is_dir_having_file, is_file_same_content};
use log::warn;

/// Same name enum as Python
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum ReplaceAction {
    Skip = 0,
    #[default]
    Replace = 1,
    Rename = 2,
    /// Check content first before deciding
    CheckReplace = 12,
}

/// Replacement strategy
#[derive(Debug, Default, Clone)]
pub struct ReplaceOptions {
    /// Strategy specified by extension
    pub ext: HashMap<String, ReplaceAction>,
    /// Default strategy
    pub default: ReplaceAction,
}

impl ReplaceOptions {
    /// Get strategy for a specific file
    fn for_path(&self, path: &Path) -> ReplaceAction {
        path.extension()
            .and_then(|s| s.to_str())
            .and_then(|ext| self.ext.get(ext).copied())
            .unwrap_or(self.default)
    }
}

/// 预设的替换策略
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, serde::Serialize, serde::Deserialize)]
pub enum ReplacePreset {
    /// 与 `ReplaceOptions::default()` 等价
    Default = 0,
    /// 与 `replace_options_update_pack()` 等价
    UpdatePack = 1,
}

impl std::str::FromStr for ReplacePreset {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "default" => Ok(ReplacePreset::Default),
            "update_pack" | "update-pack" => Ok(ReplacePreset::UpdatePack),
            _ => Err(format!(
                "Unknown preset: {}. Valid values: default, update_pack",
                s
            )),
        }
    }
}

impl clap::ValueEnum for ReplacePreset {
    fn value_variants<'a>() -> &'a [Self] {
        &[Self::Default, Self::UpdatePack]
    }

    fn to_possible_value(&self) -> Option<clap::builder::PossibleValue> {
        let name = match self {
            ReplacePreset::Default => "default",
            ReplacePreset::UpdatePack => "update_pack",
        };
        Some(clap::builder::PossibleValue::new(name))
    }
}

/// 从预设获取具体的 `ReplaceOptions`
#[must_use]
pub fn replace_options_from_preset(preset: ReplacePreset) -> ReplaceOptions {
    match preset {
        ReplacePreset::Default => ReplaceOptions::default(),
        ReplacePreset::UpdatePack => replace_options_update_pack(),
    }
}

/// Default update pack strategy
#[must_use]
pub fn replace_options_update_pack() -> ReplaceOptions {
    ReplaceOptions {
        ext: {
            BMS_FILE_EXTS
                .iter()
                .chain(BMSON_FILE_EXTS)
                .chain(&["txt"])
                .map(|ext| (ext.to_string(), ReplaceAction::CheckReplace))
                .collect()
        },
        default: ReplaceAction::Replace,
    }
}

/// Recursively move directory contents (using loops instead of recursion)
///
/// # Errors
///
/// Returns an error if file system operations fail
pub async fn move_elements_across_dir(
    dir_path_ori: impl AsRef<Path>,
    dir_path_dst: impl AsRef<Path>,
    replace_options: ReplaceOptions,
) -> io::Result<()> {
    let dir_path_ori = dir_path_ori.as_ref();
    let dir_path_dst = dir_path_dst.as_ref();
    // Lock and read source metadata
    let ori_md = match fs::metadata(&dir_path_ori).await {
        Ok(m) => m,
        Err(_) => return Ok(()),
    };

    if dir_path_ori == dir_path_dst {
        return Ok(());
    }
    if !ori_md.is_dir() {
        return Ok(());
    }
    // If target directory doesn't exist, directly move the entire directory.
    // Do this check BEFORE creating the target directory, otherwise we'd
    // unnecessarily enumerate and move children one-by-one.
    // Lock and read destination metadata
    let dst_meta_res = fs::metadata(&dir_path_dst).await;

    match dst_meta_res {
        Ok(m) => {
            if !m.is_dir() {
                return Err(io::Error::other(
                    "destination path exists and is not a directory",
                ));
            }
        }
        Err(e) => {
            if e.kind() == io::ErrorKind::NotFound {
                fs::rename(&dir_path_ori, &dir_path_dst).await?;
                return Ok(());
            }
            return Err(e);
        }
    }

    // Use queue to manage directories to be processed
    let mut pending_dirs = VecDeque::new();
    pending_dirs.push_back((dir_path_ori.to_path_buf(), dir_path_dst.to_path_buf()));

    while let Some((current_ori, current_dst)) = pending_dirs.pop_front() {
        // Process current directory with adaptive concurrency
        let next_dirs = process_directory(&current_ori, &current_dst, &replace_options).await?;

        // Add newly discovered subdirectories to the queue
        for (ori, dst) in next_dirs {
            pending_dirs.push_back((ori, dst));
        }

        // Clean up empty directories
        if (replace_options.default != ReplaceAction::Skip
            || !is_dir_having_file(&current_ori).await?)
            && let Err(e) = fs::remove_dir_all(&current_ori).await
        {
            warn!(" x PermissionError! ({}) - {}", current_ori.display(), e);
        }
    }

    Ok(())
}

/// Process a single directory, return subdirectories that need further processing
async fn process_directory(
    dir_path_ori: &Path,
    dir_path_dst: &Path,
    replace_options: &ReplaceOptions,
) -> io::Result<Vec<(PathBuf, PathBuf)>> {
    // Collect entries to be processed (files / subdirectories)
    let mut entries = fs::read_dir(dir_path_ori).await?;
    let next_folder_paths = Arc::new(Mutex::new(Vec::new()));
    let mut pairs: Vec<(PathBuf, PathBuf)> = Vec::new();

    while let Ok(Some(entry)) = entries.next_entry().await {
        let src = entry.path();
        let dst = dir_path_dst.join(entry.file_name());
        pairs.push((src, dst));
    }

    // Pre-fetch metadata concurrently (avoid Path::is_dir/is_file) with disk locks
    let metas: Vec<(
        PathBuf,
        PathBuf,
        std::fs::Metadata,
        Option<std::fs::Metadata>,
    )> = stream::iter(pairs.iter().cloned())
        .map(|(src, dst)| async move {
            let src_md = fs::metadata(&src).await?;
            let dst_md_opt = match fs::metadata(&dst).await {
                Ok(m) => Some(m),
                Err(e) if e.kind() == io::ErrorKind::NotFound => None,
                Err(e) => return Err(e),
            };
            Ok::<_, io::Error>((src, dst, src_md, dst_md_opt))
        })
        .buffer_unordered(64)
        .try_collect()
        .await?;

    // Buckets
    let mut subdir_both_exist: Vec<(PathBuf, PathBuf)> = Vec::new();
    let mut dir_direct_moves: Vec<(PathBuf, PathBuf)> = Vec::new();
    let mut file_skip_ops: Vec<(PathBuf, PathBuf)> = Vec::new();
    let mut file_rename_ops: Vec<(PathBuf, PathBuf)> = Vec::new();
    let mut file_replace_ops: Vec<(PathBuf, PathBuf)> = Vec::new();

    for (src, dst, src_md, dst_md_opt) in metas {
        if src_md.is_dir() {
            match dst_md_opt {
                Some(m) if m.is_dir() => {
                    subdir_both_exist.push((src, dst));
                }
                _ => {
                    // Destination missing or not a directory -> move directly
                    dir_direct_moves.push((src, dst));
                }
            }
        } else if src_md.is_file() {
            let action = replace_options.for_path(&src);
            match action {
                ReplaceAction::Skip => file_skip_ops.push((src, dst)),
                ReplaceAction::Rename => file_rename_ops.push((src, dst)),
                _ => file_replace_ops.push((src, dst)),
            }
        }
    }

    // Stage 1: both side subdirectories exist -> enqueue for next round
    {
        let mut next = next_folder_paths.lock().await;
        next.extend(subdir_both_exist);
    }

    // Stage 2a: directory direct moves (streamed parallel)
    stream::iter(dir_direct_moves)
        .map(|(src, dst)| async move { fs::rename(&src, &dst).await.map(|_| ()) })
        .buffer_unordered(64)
        .try_for_each(|_| async { Ok(()) })
        .await?;

    // Stage 2b: file Skip actions (streamed parallel)
    let rep_clone = replace_options.clone();
    stream::iter(file_skip_ops)
        .map(|(src, dst)| {
            let rep = rep_clone.clone();
            async move {
                let exists = fs::metadata(&dst).await.is_ok();
                if exists {
                    return Ok::<(), io::Error>(());
                }
                move_file(&src, &dst, &rep).await.map(|_| ())
            }
        })
        .buffer_unordered(128)
        .try_for_each(|_| async { Ok(()) })
        .await?;

    // Stage 2c: file Rename actions (streamed parallel)
    stream::iter(file_rename_ops)
        .map(|(src, dst)| async move { move_file_rename(&src, &dst).await.map(|_| ()) })
        .buffer_unordered(128)
        .try_for_each(|_| async { Ok(()) })
        .await?;

    // Stage 3: remaining overwrites (Replace / CheckReplace) (streamed parallel)
    let rep_clone2 = replace_options.clone();
    stream::iter(file_replace_ops)
        .map(|(src, dst)| {
            let rep = rep_clone2.clone();
            async move { move_file(&src, &dst, &rep).await.map(|_| ()) }
        })
        .buffer_unordered(128)
        .try_for_each(|_| async { Ok(()) })
        .await?;

    // Return subdirectories that need further processing
    Ok(next_folder_paths.lock().await.clone())
}

// removed unused move_action

/// Move a single file, handle conflicts according to strategy
async fn move_file(src: &Path, dst: &Path, rep: &ReplaceOptions) -> io::Result<()> {
    let action = rep.for_path(src);

    match action {
        ReplaceAction::Replace => fs::rename(src, dst).await,
        ReplaceAction::Skip => {
            let exists = fs::metadata(&dst).await.is_ok();
            if exists {
                return Ok(());
            }
            fs::rename(src, dst).await
        }
        ReplaceAction::Rename => move_file_rename(src, dst).await,
        ReplaceAction::CheckReplace => {
            let dst_exists = fs::metadata(&dst).await.is_ok();
            if !dst_exists {
                fs::rename(src, dst).await
            } else {
                let same = is_file_same_content(src, dst).await?;
                if same {
                    // Same content, directly overwrite
                    fs::rename(src, dst).await
                } else {
                    move_file_rename(src, dst).await
                }
            }
        }
    }
}

/// "Rename" move with retry
async fn move_file_rename(src: &Path, dst_dir: &Path) -> io::Result<()> {
    let mut dst = dst_dir.to_path_buf();
    let stem = src.file_stem().and_then(|s| s.to_str()).unwrap_or("file");
    let ext = src.extension().and_then(|s| s.to_str()).unwrap_or("");

    let mut _count = 0;
    for i in std::iter::from_fn(|| {
        _count += 1;
        Some(_count)
    }) {
        let name = if i == 0 {
            format!("{stem}.{ext}")
        } else {
            format!("{stem}.{i}.{ext}")
        };
        dst.set_file_name(name);
        if fs::metadata(&dst).await.is_err() {
            fs::rename(src, &dst).await?;
            return Ok(());
        }
        let same = is_file_same_content(src, &dst).await?;
        if same {
            // File with same name and content already exists, skip
            fs::remove_file(src).await?;
            return Ok(());
        }
    }
    Err(io::Error::other("too many duplicate files"))
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::{TempDir, tempdir};
    use tokio::{fs, io};

    /// Create test directory structure
    async fn create_test_structure(base_dir: &Path) -> io::Result<()> {
        // Create subdirectory
        let sub_dir = base_dir.join("subdir");
        fs::create_dir_all(&sub_dir).await?;

        // Create files
        fs::write(base_dir.join("file1.txt"), "content1").await?;
        fs::write(base_dir.join("file2.bms"), "content2").await?;
        fs::write(sub_dir.join("file3.txt"), "content3").await?;

        // Create nested directory
        let nested_dir = sub_dir.join("nested");
        fs::create_dir_all(&nested_dir).await?;
        fs::write(nested_dir.join("file4.txt"), "content4").await?;

        Ok(())
    }

    /// Verify directory structure
    async fn verify_structure(dir: &Path, expected_files: &[(&str, &str)]) -> io::Result<()> {
        for (file_path, expected_content) in expected_files {
            let full_path = dir.join(file_path);
            assert!(
                full_path.exists(),
                "File does not exist: {}",
                full_path.display()
            );

            let content = fs::read_to_string(&full_path).await?;
            assert_eq!(
                &content,
                expected_content,
                "File content mismatch: {}",
                full_path.display()
            );
        }
        Ok(())
    }

    /// Clean up test directory
    async fn cleanup_test_dir(dir: &TempDir) {
        if let Err(e) = fs::remove_dir_all(dir.path()).await {
            eprintln!("Failed to clean up test directory: {e}");
        }
    }

    #[tokio::test]
    async fn test_move_elements_across_dir_basic() {
        let temp_dir = tempdir().expect("Failed to create temp directory");
        let src_dir = temp_dir.path().join("src");
        let dst_dir = temp_dir.path().join("dst");

        // Create source directory structure
        fs::create_dir_all(&src_dir)
            .await
            .expect("Failed to create source directory");
        create_test_structure(&src_dir)
            .await
            .expect("Failed to create test structure");

        // Execute move
        let replace_options = ReplaceOptions::default();

        move_elements_across_dir(&src_dir, &dst_dir, replace_options)
            .await
            .expect("Move operation failed");

        // Verify result
        let expected_files = [
            ("file1.txt", "content1"),
            ("file2.bms", "content2"),
            ("subdir/file3.txt", "content3"),
            ("subdir/nested/file4.txt", "content4"),
        ];

        verify_structure(&dst_dir, &expected_files)
            .await
            .expect("Failed to verify structure");

        // Verify source directory has been cleaned up
        assert!(!src_dir.exists(), "Source directory should be deleted");

        cleanup_test_dir(&temp_dir).await;
    }

    #[tokio::test]
    async fn test_move_elements_across_dir_skip_existing() {
        let temp_dir = tempdir().expect("Failed to create temp directory");
        let src_dir = temp_dir.path().join("src");
        let dst_dir = temp_dir.path().join("dst");

        // Create source directory structure
        fs::create_dir_all(&src_dir)
            .await
            .expect("Failed to create source directory");
        create_test_structure(&src_dir)
            .await
            .expect("Failed to create test structure");

        // Create file with same name in target directory
        fs::create_dir_all(&dst_dir)
            .await
            .expect("Failed to create target directory");
        fs::write(dst_dir.join("file1.txt"), "existing_content")
            .await
            .expect("Failed to create file");

        // Use Skip strategy
        let replace_options = ReplaceOptions {
            default: ReplaceAction::Skip,
            ..Default::default()
        };

        move_elements_across_dir(&src_dir, &dst_dir, replace_options)
            .await
            .expect("Move operation failed");

        // Verify target file keeps original content
        let content = fs::read_to_string(dst_dir.join("file1.txt"))
            .await
            .expect("Failed to read file");
        assert_eq!(
            content, "existing_content",
            "File content should remain unchanged"
        );

        // Verify other files were moved
        assert!(dst_dir.join("file2.bms").exists());
        assert!(dst_dir.join("subdir").exists());

        cleanup_test_dir(&temp_dir).await;
    }

    #[tokio::test]
    async fn test_move_elements_across_dir_rename_conflict() {
        let temp_dir = tempdir().expect("Failed to create temp directory");
        let src_dir = temp_dir.path().join("src");
        let dst_dir = temp_dir.path().join("dst");

        // Create source directory structure
        fs::create_dir_all(&src_dir)
            .await
            .expect("Failed to create source directory");
        fs::write(src_dir.join("file1.txt"), "new_content")
            .await
            .expect("Failed to create file");

        // Create file with same name in target directory
        fs::create_dir_all(&dst_dir)
            .await
            .expect("Failed to create target directory");
        fs::write(dst_dir.join("file1.txt"), "existing_content")
            .await
            .expect("Failed to create file");

        // Use Rename strategy
        let replace_options = ReplaceOptions {
            default: ReplaceAction::Rename,
            ..Default::default()
        };

        move_elements_across_dir(&src_dir, &dst_dir, replace_options)
            .await
            .expect("Move operation failed");

        // Verify original file exists
        let content = fs::read_to_string(dst_dir.join("file1.txt"))
            .await
            .expect("Failed to read file");
        assert_eq!(
            content, "existing_content",
            "Original file should remain unchanged"
        );

        // Verify new file was renamed
        assert!(
            dst_dir.join("file1.1.txt").exists(),
            "Should create renamed file"
        );

        cleanup_test_dir(&temp_dir).await;
    }

    #[tokio::test]
    async fn test_move_elements_across_dir_check_replace() {
        let temp_dir = tempdir().expect("Failed to create temp directory");
        let src_dir = temp_dir.path().join("src");
        let dst_dir = temp_dir.path().join("dst");

        // Create source directory structure
        fs::create_dir_all(&src_dir)
            .await
            .expect("Failed to create source directory");
        fs::write(src_dir.join("file1.txt"), "same_content")
            .await
            .expect("Failed to create file");

        // Create file with same name in target directory, same content
        fs::create_dir_all(&dst_dir)
            .await
            .expect("Failed to create target directory");
        fs::write(dst_dir.join("file1.txt"), "same_content")
            .await
            .expect("Failed to create file");

        // Use CheckReplace strategy
        let replace_options = ReplaceOptions {
            default: ReplaceAction::CheckReplace,
            ..Default::default()
        };

        move_elements_across_dir(&src_dir, &dst_dir, replace_options)
            .await
            .expect("Move operation failed");

        // Verify file was overwritten (because content is the same)
        let content = fs::read_to_string(dst_dir.join("file1.txt"))
            .await
            .expect("Failed to read file");
        assert_eq!(
            content, "same_content",
            "File content should remain unchanged"
        );

        // Verify source directory was cleaned up
        assert!(!src_dir.exists(), "Source directory should be deleted");

        cleanup_test_dir(&temp_dir).await;
    }

    #[tokio::test]
    async fn test_move_elements_across_dir_same_directory() {
        let temp_dir = tempdir().expect("Failed to create temp directory");
        let src_dir = temp_dir.path().join("src");

        // Create source directory structure
        fs::create_dir_all(&src_dir)
            .await
            .expect("Failed to create source directory");
        create_test_structure(&src_dir)
            .await
            .expect("Failed to create test structure");

        // Try to move to the same directory
        let replace_options = ReplaceOptions::default();
        let result = move_elements_across_dir(&src_dir, &src_dir, replace_options).await;
        assert!(result.is_ok(), "Moving to same directory should succeed");

        // Verify directory structure remains unchanged
        assert!(src_dir.exists(), "Source directory should still exist");
        assert!(
            src_dir.join("file1.txt").exists(),
            "File should still exist"
        );

        cleanup_test_dir(&temp_dir).await;
    }

    #[tokio::test]
    async fn test_move_elements_across_dir_nonexistent_source() {
        let temp_dir = tempdir().expect("Failed to create temp directory");
        let src_dir = temp_dir.path().join("nonexistent");
        let dst_dir = temp_dir.path().join("dst");

        let replace_options = ReplaceOptions::default();
        let result = move_elements_across_dir(&src_dir, &dst_dir, replace_options).await;
        assert!(
            result.is_ok(),
            "Moving non-existent directory should succeed (no operation)"
        );

        cleanup_test_dir(&temp_dir).await;
    }

    #[tokio::test]
    async fn test_move_elements_across_dir_nonexistent_target() {
        let temp_dir = tempdir().expect("Failed to create temp directory");
        let src_dir = temp_dir.path().join("src");
        let dst_dir = temp_dir.path().join("nonexistent_dst");

        // Create source directory structure
        fs::create_dir_all(&src_dir)
            .await
            .expect("Failed to create source directory");
        create_test_structure(&src_dir)
            .await
            .expect("Failed to create test structure");

        let replace_options = ReplaceOptions::default();
        let result = move_elements_across_dir(&src_dir, &dst_dir, replace_options).await;
        assert!(
            result.is_ok(),
            "Moving to non-existent target should succeed"
        );

        // Verify the entire directory was moved (renamed)
        assert!(!src_dir.exists(), "Source directory should not exist");
        assert!(dst_dir.exists(), "Target directory should exist");

        // Verify all files were moved
        let expected_files = [
            ("file1.txt", "content1"),
            ("file2.bms", "content2"),
            ("subdir/file3.txt", "content3"),
            ("subdir/nested/file4.txt", "content4"),
        ];
        verify_structure(&dst_dir, &expected_files)
            .await
            .expect("Failed to verify structure");

        cleanup_test_dir(&temp_dir).await;
    }

    #[tokio::test]
    async fn test_move_elements_across_dir_with_ext_specific_rules() {
        let temp_dir = tempdir().expect("Failed to create temp directory");
        let src_dir = temp_dir.path().join("src");
        let dst_dir = temp_dir.path().join("dst");

        // Create source directory structure
        fs::create_dir_all(&src_dir)
            .await
            .expect("Failed to create source directory");
        fs::write(src_dir.join("file1.txt"), "content1")
            .await
            .expect("Failed to create file");
        fs::write(src_dir.join("file2.bms"), "content2")
            .await
            .expect("Failed to create file");
        fs::write(src_dir.join("file3.other"), "content3")
            .await
            .expect("Failed to create file");

        // Create conflicting files in target directory
        fs::create_dir_all(&dst_dir)
            .await
            .expect("Failed to create target directory");
        fs::write(dst_dir.join("file1.txt"), "existing_txt")
            .await
            .expect("Failed to create file");
        fs::write(dst_dir.join("file2.bms"), "existing_bms")
            .await
            .expect("Failed to create file");
        fs::write(dst_dir.join("file3.other"), "existing_other")
            .await
            .expect("Failed to create file");

        // Use specific extension rules
        let mut replace_options = ReplaceOptions::default();
        replace_options
            .ext
            .insert("txt".to_string(), ReplaceAction::Skip);
        replace_options
            .ext
            .insert("bms".to_string(), ReplaceAction::Rename);
        replace_options.default = ReplaceAction::Replace;

        move_elements_across_dir(&src_dir, &dst_dir, replace_options)
            .await
            .expect("Move operation failed");

        // Verify txt file was skipped
        let content = fs::read_to_string(dst_dir.join("file1.txt"))
            .await
            .expect("Failed to read file");
        assert_eq!(content, "existing_txt", "txt file should be skipped");

        // Verify bms file was renamed
        assert!(
            dst_dir.join("file2.1.bms").exists(),
            "bms file should be renamed"
        );

        // Verify other file was replaced
        let other_content = fs::read_to_string(dst_dir.join("file3.other"))
            .await
            .expect("Failed to read file");
        assert_eq!(other_content, "content3", "other file should be replaced");

        cleanup_test_dir(&temp_dir).await;
    }
}

File: src/lib/utils/fs/moving.ts
/**
 * 跨目录文件移动工具（含冲突处理）
 */

import { exists, readDir, remove, rename, stat } from '@tauri-apps/plugin-fs';
import { isDirHavingFile, isFileSameContent } from './compare';

/**
 * 替换操作类型
 */
export enum ReplaceAction {
  Skip = 0,
  Replace = 1,
  Rename = 2,
  CheckReplace = 12,
}

/**
 * 替换选项
 */
export interface ReplaceOptions {
  /** 按扩展名指定的策略 */
  ext: Record<string, ReplaceAction>;
  /** 默认策略 */
  default: ReplaceAction;
}

/**
 * 替换预设
 */
export enum ReplacePreset {
  Default = 0,
  UpdatePack = 1,
}

/**
 * 从预设获取替换选项
 */
export function replaceOptionsFromPreset(preset: ReplacePreset): ReplaceOptions {
  switch (preset) {
    case ReplacePreset.Default:
      return { ext: {}, default: ReplaceAction.Replace };
    case ReplacePreset.UpdatePack:
      return replaceOptionsUpdatePack();
  }
}

/**
 * 默认的更新包策略
 */
function replaceOptionsUpdatePack(): ReplaceOptions {
  return {
    ext: {
      bms: ReplaceAction.CheckReplace,
      bme: ReplaceAction.CheckReplace,
      bml: ReplaceAction.CheckReplace,
      pms: ReplaceAction.CheckReplace,
      bmson: ReplaceAction.CheckReplace,
      txt: ReplaceAction.CheckReplace,
    },
    default: ReplaceAction.Replace,
  };
}

/**
 * 获取特定文件的替换策略
 */
async function getActionForPath(options: ReplaceOptions, filePath: string): Promise<ReplaceAction> {
  const { getFileExtension } = await import('./path');
  const ext = getFileExtension(filePath);
  return options.ext[ext] ?? options.default;
}

/**
 * 递归移动目录内容
 */
export async function moveElementsAcrossDir(
  fromDir: string,
  toDir: string,
  replaceOptions: ReplaceOptions
): Promise<void> {
  // 检查源目录元数据
  let fromMeta;
  try {
    fromMeta = await stat(fromDir);
  } catch {
    // 源目录不存在，直接返回
    return;
  }

  // 源和目标相同，跳过
  if (fromDir === toDir) {
    return;
  }

  // 源不是目录，跳过
  if (!fromMeta.isDirectory) {
    return;
  }

  // 检查目标目录
  const toExists = await exists(toDir);

  if (!toExists) {
    // 目标不存在，直接重命名整个目录
    await rename(fromDir, toDir);
    return;
  }

  // 目标存在但不是目录
  const toMeta = await stat(toDir);
  if (!toMeta.isDirectory) {
    throw new Error('目标路径存在且不是目录');
  }

  // 使用队列管理待处理的目录
  const queue: [string, string][] = [[fromDir, toDir]];

  while (queue.length > 0) {
    const [currentFrom, currentTo] = queue.shift()!;

    // 处理当前目录
    const subdirs = await processDirectory(currentFrom, currentTo, replaceOptions);

    // 将发现的子目录加入队列
    queue.push(...subdirs);

    // 清理空目录
    const hasFiles = await isDirHavingFile(currentFrom);
    const skipCleanup = replaceOptions.default === ReplaceAction.Skip && hasFiles;

    if (!skipCleanup) {
      try {
        await remove(currentFrom, { recursive: true });
      } catch (error) {
        console.warn(`权限不足，无法删除 ${currentFrom}:`, error);
      }
    }
  }
}

/**
 * 处理单个目录，返回需要进一步处理的子目录
 */
async function processDirectory(
  fromDir: string,
  toDir: string,
  replaceOptions: ReplaceOptions
): Promise<[string, string][]> {
  const subdirs: [string, string][] = [];

  // 收集目录条目
  const entries = await readDir(fromDir);
  const pairs: Array<{ src: string; dst: string; name: string }> = [];

  for (const entry of entries) {
    if (!entry.name) continue;

    pairs.push({
      src: `${fromDir}/${entry.name}`,
      dst: `${toDir}/${entry.name}`,
      name: entry.name,
    });
  }

  // 预取元数据（并发）
  const metas = await Promise.all(
    pairs.map(async (pair) => {
      try {
        const srcMeta = await stat(pair.src);
        let dstMeta;
        try {
          dstMeta = await stat(pair.dst);
        } catch {
          dstMeta = null;
        }

        return { ...pair, srcMeta, dstMeta };
      } catch {
        return { ...pair, srcMeta: null, dstMeta: null };
      }
    })
  );

  // 分类
  const subdirBothExist: [string, string][] = [];
  const dirDirectMoves: [string, string][] = [];
  const fileSkipOps: Array<{ src: string; dst: string }> = [];
  const fileRenameOps: Array<{ src: string; dst: string }> = [];
  const fileReplaceOps: Array<{ src: string; dst: string }> = [];

  for (const { src, dst, srcMeta, dstMeta } of metas) {
    if (!srcMeta) continue;

    if (srcMeta.isDirectory) {
      if (dstMeta && dstMeta.isDirectory) {
        subdirBothExist.push([src, dst]);
      } else {
        dirDirectMoves.push([src, dst]);
      }
    } else if (srcMeta.isFile) {
      const action = await getActionForPath(replaceOptions, src);

      switch (action) {
        case ReplaceAction.Skip:
          fileSkipOps.push({ src, dst });
          break;
        case ReplaceAction.Rename:
          fileRenameOps.push({ src, dst });
          break;
        default:
          fileReplaceOps.push({ src, dst });
          break;
      }
    }
  }

  // Stage 1: 双边都存在的子目录 -> 加入队列
  subdirs.push(...subdirBothExist);

  // Stage 2a: 目录直接移动（并发）
  await Promise.all(
    dirDirectMoves.map(async ([src, dst]) => {
      try {
        await rename(src, dst);
      } catch (error) {
        console.warn(`Failed to move directory ${src} -> ${dst}:`, error);
      }
    })
  );

  // Stage 2b: 文件 Skip 操作（并发）
  await Promise.all(
    fileSkipOps.map(async ({ src, dst }) => {
      const dstExists = await exists(dst);
      if (dstExists) {
        return; // 跳过
      }
      await moveFile(src, dst, replaceOptions);
    })
  );

  // Stage 2c: 文件 Rename 操作（并发）
  await Promise.all(
    fileRenameOps.map(async ({ src, dst }) => {
      await moveFileRename(src, dst);
    })
  );

  // Stage 3: 剩余的覆盖操作（Replace / CheckReplace）（并发）
  await Promise.all(
    fileReplaceOps.map(async ({ src, dst }) => {
      await moveFile(src, dst, replaceOptions);
    })
  );

  return subdirs;
}

/**
 * 移动单个文件，根据策略处理冲突
 */
async function moveFile(src: string, dst: string, options: ReplaceOptions): Promise<void> {
  const action = await getActionForPath(options, src);

  switch (action) {
    case ReplaceAction.Replace:
      await rename(src, dst);
      break;

    case ReplaceAction.Skip:
      if (await exists(dst)) {
        return; // 跳过
      }
      await rename(src, dst);
      break;

    case ReplaceAction.Rename:
      await moveFileRename(src, dst);
      break;

    case ReplaceAction.CheckReplace:
      if (await exists(dst)) {
        const same = await isFileSameContent(src, dst);
        if (same) {
          // 内容相同，直接覆盖
          await rename(src, dst);
        } else {
          // 内容不同，重命名
          await moveFileRename(src, dst);
        }
      } else {
        await rename(src, dst);
      }
      break;
  }
}

/**
 * 重命名移动文件（带重试）
 */
async function moveFileRename(src: string, dstDir: string): Promise<void> {
  const srcFileName = src.split('/').pop() || src.split('\\').pop() || 'file';
  const { getFileStem, getFileExtension } = await import('./path');

  const stem = getFileStem(srcFileName);
  const ext = getFileExtension(srcFileName);

  let count = 0;
  while (true) {
    count++;

    const newName = count === 1 ? `${stem}.${ext}` : `${stem}.${count}.${ext}`;
    const dst = `${dstDir.split('/').slice(0, -1).join('/')}/${newName}`;

    if (!(await exists(dst))) {
      await rename(src, dst);
      return;
    }

    // 检查内容是否相同
    const same = await isFileSameContent(src, dst);
    if (same) {
      // 文件已存在且内容相同，删除源文件
      await remove(src);
      return;
    }

    if (count > 1000) {
      throw new Error('重复文件过多');
    }
  }
}
